quinn/src/platform/unix.rs

```diff
@@ -71,51 +71,27 @@ impl super::UdpExt for UdpSocket {
         Ok(())
     }
 
-    fn send_ext(
-        &self,
-        remote: &SocketAddr,
-        ecn: Option<EcnCodepoint>,
-        msg: &[u8],
-    ) -> io::Result<usize> {
-        let (name, namelen) = match *remote {
-            SocketAddr::V4(ref addr) => {
-                (addr as *const _ as _, mem::size_of::<libc::sockaddr_in>())
-            }
-            SocketAddr::V6(ref addr) => {
-                (addr as *const _ as _, mem::size_of::<libc::sockaddr_in6>())
-            }
-        };
-        let ecn = ecn.map_or(0, |x| x as libc::c_int);
-        let mut iov = libc::iovec {
-            iov_base: msg.as_ptr() as *const _ as *mut _,
-            iov_len: msg.len(),
-        };
-        let mut hdr: libc::msghdr = unsafe { mem::zeroed() };
-        hdr.msg_name = name;
-        hdr.msg_namelen = namelen as _;
-        hdr.msg_iov = &mut iov;
-        hdr.msg_iovlen = 1;
-        hdr.msg_control = ptr::null_mut();
-        hdr.msg_controllen = 0;
-        hdr.msg_flags = 0;
-        // We may never fully initialize this, and it's only written/read via `ptr::write`/syscalls,
-        // so no `assume_init` call can or should be made.
-        let mut ctrl = cmsg::Aligned(MaybeUninit::<[u8; CMSG_LEN]>::uninit());
-        hdr.msg_control = ctrl.0.as_mut_ptr() as _;
-        hdr.msg_controllen = CMSG_LEN as _;
-        let is_ipv4 = match remote {
-            SocketAddr::V4(_) => true,
-            SocketAddr::V6(ref addr) => addr.ip().segments().starts_with(&[0, 0, 0, 0, 0, 0xffff]),
-        };
-        let mut encoder = unsafe { cmsg::Encoder::new(&mut hdr) };
-        if is_ipv4 {
-            encoder.push(libc::IPPROTO_IP, libc::IP_TOS, ecn as IpTosTy);
-        } else {
-            encoder.push(libc::IPPROTO_IPV6, libc::IPV6_TCLASS, ecn);
+    #[cfg(not(target_os = "macos"))]
+    fn send_ext(&self, transmits: &[Transmit]) -> io::Result<usize> {
+        use crate::udp::BATCH_SIZE;
+        let mut msgs: [libc::mmsghdr; BATCH_SIZE] = unsafe { mem::zeroed() };
+        let mut iovecs: [libc::iovec; BATCH_SIZE] = unsafe { mem::zeroed() };
+        let mut cmsgs = [cmsg::Aligned(MaybeUninit::uninit()); BATCH_SIZE];
+        for (hdr, (iov, (ctrl, transmit))) in msgs
+            .iter_mut()
+            .zip(iovecs.iter_mut().zip(cmsgs.iter_mut().zip(transmits)))
```

Comment:
Isn't there some `flatten()` thing we can use here?

Reply:
Not that I know if. `itertools` has `zip4` and such but this doesn't seem bad enough to justify a dep.

Reply:
On second thought, maybe it would be better to use explicit indexing here?

Reply:
Applied.